features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,402 @@
+# auth gear and Cloud Function interaction overview in skygear next
+
+Base on the new product architecture decision (auth gear + Cloud Function and drop record gear), skygear next should bring up a new experience on how auth gear is used in Cloud Function and Client SDK.
+
+## Glossary
+
+* auth gear: a skygear provided component which utilizes user authentication and authorization process, and user profile handling. It would bring enhanced features in the future, such as: JWT provider, Auth UIKit, user management dashboard.
+* Cloud Function: a developer would create a Cloud Function to fulfill application requirements. A cloud function should be a single purpose that attached to certain events or triggered by requirement.
+* user auth data: such as disabled, last login at, ..., etc.
+* user common properties: such as avatar, first name, last name, ..., etc.
+* user profile: varied user properties, differs from application to application.
+* user attributes: merge user auth data, user common properties, and user profile together.
+
+## Goal
+
+* Consider auth gear would be extended to be a rich-featured functional set, such as JWT provider, Auth UIKit, ..., etc.
+* Encourage Cloud Functions, less SDK burden, a developer should create corresponding Cloud Function to fulfill application requirements.
+* Avoid coupling between cloud function and auth gear, consider auth gear would be used independently, any access to data owned by another service should only happen through APIs.
+* Auth gear and Cloud Function both should have current user execution context.
+* List use cases to demonstrate how auth gear and Cloud Function interaction flow.
+
+## Major changes
+
+* Move admin related features from Client SDK to APIs at Cloud Functions.
+* API gateway should inject "current user" into request context and then dispatch request to auth gear or Cloud Function.
+* Auth gear should extend core user to support common user properties.
+
+## Architecture overview
+
+![](https://i.imgur.com/jxNxhmd.png)
+
+\* this diagram is an overview only, detail flow explain in later section.
+
+- Cloud Functions may have its own DB.
+- It's developer's responsibility to create corresponding Cloud Function to handle admin related auth features (enable/disable user, assign role, ...).
+- Cloud Function would implement hook functions to handle user changes.
+
+## Move auth admin APIs from SDK to Cloud Function
+
+Following function will be removed from SDK, a developer should create its Cloud Function to handle admin tasks.
+
+| Attribute | Interface | Old JS SDK API |
+| -------- | -------- | ------ |
+| disabled | `POST /auth/disable/set`| `adminDisableUser`<br/>`adminEnableUser` |
+| password | `POST /auth/reset_password`| `adminResetPassword` |
+| role | `POST /auth/role/assign`<br/>`POST /auth/role/admin`<br/>`POST /auth/role/default` | `assignUserRole`<br/>`revokeUserRole`<br/>`setDefaultRole` |
```

Comment:
Shall we add the APIs for update user profile and metadata too?

Reply:
I don't think so, `update_user_meta_data_and_profile`(temp name) API is not admin API, it will need client SDK support. This section lists API supports **removed** from client SDK (JS, iOS, Android).